docs/docs/reference/other-new-features/explicit-nulls.md

```diff
@@ -0,0 +1,296 @@
+---
+layout: doc-page
+title: "Explicit Nulls"
+---
+
+This proposal describes a modification to the Scala type system that makes reference types
+(anything that extends `AnyRef`) _non-nullable_.
+
+This means the following code will no longer typecheck:
+```
+val x: String = null // error: found `Null`,  but required `String`
+```
+
+Instead, to mark a type as nullable we use a [type union](https://dotty.epfl.ch/docs/reference/new-types/union-types.html)
+```
+val x: String|Null = null // ok
+```
+
+Explicit nulls are enabled via a `-Yexplicit-nulls` flag, so they're an opt-in feature.
+
+Read on for details.
+
+## New Type Hierarchy
+
+When explicit nulls are enabled, the type hierarchy changes so that `Null` is subtype only of
+`Any`, as opposed to every reference type.
+
+This is the new type hierarchy:
+![](../../images/explicit-nulls/explicit-nulls-type-hierarchy.png "Type Hierarchy for Explicit Nulls")
+
+After erasure, `Null` remains a subtype of all reference types (as forced by the JVM).
+
+## Unsoundness
+
+The new type system is unsound with respect to `null`. This means there are still instances where an expressions has a non-nullable type like `String`, but its value is `null`.
+
+The unsoundness happens because uninitialized fields in a class start out as `null`:
+```scala
+class C {
+  val f: String = foo(f)
+  def foo(f2: String): String = if (f2 == null) "field is null" else f2
+}
+val c = new C()
+// c.f == "field is null"
+```
+
+Enforcing sound initialization is a non-goal of this proposal. However, once we have a type
+system where nullability is explicit, we can use a sound initialization scheme like the one
+proposed by @liufengyun and @biboudis in [https://github.com/lampepfl/dotty/pull/4543](https://github.com/lampepfl/dotty/pull/4543) to eliminate this particular source of unsoundness.
+
+## Equality
+
+Because of the unsoundness, we need to allow comparisons of the form `x == null` or `x != null`
+even when `x` has a non-nullable reference type (but not a value type). This is so we have an
+"escape hatch" for when we know `x` is nullable even when the type says it shouldn't be.
+```scala
+val x: String|Null = null
+x == null       // ok: x is a nullable string
+"hello" == null // ok: String is a reference type
+1 == null       // error: Int is a value type
+```
+
+### Reference Equality
+
+Recall that `Null` is now a direct subtype of `Any`, as opposed to `AnyRef`.
+However, we also need to allow reference equality comparisons:
+```scala
+val x: String = null
+x eq null // ok: could return `true` because of unsoundness
+```
+
+We support this case by making the `eq` and `ne` methods in `AnyRef` take a `AnyRef|Null`
+as argument.
+
+We also need to support
+```scala
+null.eq("hello")
+val x: String|Null = null
+x.eq(null)
+```
+
+We support this case via extension methods defined in the Predef:
+```scala
+def (x: AnyRef|Null) eq(y: AnyRef|Null): Boolean =
+  (x == null && y == null) || (x != null && x.eq(y))
```

Comment:
Without flow typing, the last call to `x.eq(y)` will call back into this extension method, causing an infinite recursion, won't it? Even *with* flow typing, I would find this fragile. You should explicitly cast both `x` and `y` too `AnyRef` in that last comparison.

And since you need the casts anyway, you can take advantage of the type erasure to implement the whole method as just
```scala
... = x.asInstanceOf[AnyRef] eq y.asInstanceOf[AnyRef]
```